h/settings.py

```diff
@@ -13,54 +14,90 @@ class SettingError(Exception):
     pass
 
 
-class DeprecatedSetting(object):
-    """A wrapper for deprecated settings which emits appropriate warnings."""
-
-    def __init__(self, setting, message):
-        self.setting = setting
-        self.message = message
-
-        # Test seam
-        self.warn = log.warn
-
-    def __call__(self, environ):
-        result = self.setting(environ)
-        if result is not None:
-            self.warn(self.warning)
-        return result
-
-    @property
-    def warning(self):
-        return 'use of {s} is deprecated: {m}'.format(s=self.setting,
-                                                      m=self.message)
-
-
-class EnvSetting(object):
-
-    """An (optionally typed) setting read from an environment variable."""
-
-    def __init__(self, setting, varname, type=None):
-        self.setting = setting
-        self.varname = varname
-        if type is not None:
-            self.type = type
-        else:
-            self.type = str
-
-    def __call__(self, environ):
-        if self.varname in environ:
+class SettingsManager(object):
+    """
+    Configuration setting resolver.
+
+    SettingsManager resolves settings from various sources into the final typed
+    values used when the app runs. It also provides a way to check for missing
+    required settings or use of deprecated settings.
+
+    The resolved settings are available via the `settings` attribute.
+    """
+
+    def __init__(self, settings={}, environ=None):
+        """
+        Initialize SettingsManager with initial setting values from config files
+        and the environment.
+
+        :param settings: Initial configuration settings read from config files
+        :type settings: Dict[str,str]
+        :param environ: Environment variable mappings
+        :type environ: Dict[str, str]
+        """
+
+        if environ is None:
+            environ = os.environ
+
+        self.settings = {}
+        self.settings.update(settings)
+
+        self._environ = environ
+
+    def set(self,  # pylint: disable=too-many-arguments
+            name,
+            envvar,
+            type_=str,
+            required=False,
+            default=None,
+            deprecated_msg=None):
+        """
+        Resolve a typed value for a setting.
+
+        Sets the setting `name` to the value from the highest priority source
+        and coerces the result to a typed value using `type_`. The sources
+        in priority order are: 1) Environment variables (`envvar`), 2)
+        Configuration files (passed to the constructor), 3) Defaults (`default`).
```

Comment:
I wonder if just the first two sentences from my suggested docstring above (plus the `:param`s) would be enough...

Reply:
> Set settings[name] to the right value.
> 
> Set `settings[name]` to `environ[envvar]`, or `settings[name]` (i.e. leave it unchanged), or `default`, or leave it unset, in that order. `environ` overrides `settings`, and `settings` overrides `default`).

This seems pretty concise and concrete. The `:param`s for `required`, `type_` and `deprecated_msg` can then explain what those do, they don't need to be in the main docstring.

Reply:
> Set `settings[name]` to `environ[envvar]`

This is not exactly accurate because of the type-casting that happens. I've revised the description along these lines though.